pcs_and_grm.R

```diff
@@ -0,0 +1,75 @@
+#! /usr/bin/env Rscript
+library(argparser)
+library(SeqArray)
+library(GENESIS)
+library(SeqVarTools)
+library(magrittr)
+library(SNPRelate)
+
+sessionInfo()
+
+# read arguments
+argp <- arg_parser("Generate PCs and GRM") %>%
+  add_argument("gds_file", help = "GDS file") %>%
+  add_argument("--out_prefix", help = "Prefix for output files") %>%
+  add_argument("--variant_id", help = "File with vector of variant IDs") %>%
+  add_argument("--sample_id", help = "File with vector of sample IDs") %>%
+  add_argument("--kin.thresh", default = 5.5,
+               help = "Kinship threshold for pcair (2 ^ -kin.thresh)") %>%
+  add_argument("--n_pcs", default = 3,
+               "Number of PCs to pass to pcrelate") %>%
+  add_argument("--keep_king", flag = TRUE, help = "Save KING-robust GRM")
+
+argv <- parse_args(argp)
+if (!is.na(argv$variant_id)) {
+  variant_id <- readRDS(argv$variant_id)
+} else {
+  variant_id <- NULL
+}
+if (!is.na(argv$variant_id)) {
+  sample_id <- readRDS(argv$sample_id)
+} else {
+  sample_id <- NULL
+}
+kin.thresh <- 2 ^ (-argv$kin.thresh)
+out_prefix <- ifelse(!is.na(argv$out_prefix), argv$out_prefix, "")
+gds <- seqOpen(argv$gds_file)
+
+king <- snpgdsIBDKING(gds, snp.id = variant_id, sample.id = sample_id)
+kingMat <- king$kinship
+colnames(kingMat) <- rownames(kingMat) <- king$sample.id
+
+if (argv$keep_king) {
+  kingMat_temp <- kingMat * 2 # Scaled to match pc-relate GRM
+  # coerces low values in matrix to 0
+  kingMat_temp[kingMat_temp <= kin.thresh] <- 0
+  saveRDS(kingMat_temp, paste0(out_prefix, "king_robust_grm.rds"))
+  rm(kingMat_temp)
+}
+
+mypcair <- pcair(gds, kinobj = kingMat, kin.thresh = kin.thresh,
+                 divobj = kingMat, snp.include = variant_id,
+                 sample.include = sample_id)
+
+seqSetFilter(gds, variant.id = variant_id, sample.id = sample_id)
+seqData <- SeqVarData(gds)
+print("1st iteration PC-relate")
+iterator <- SeqVarBlockIterator(seqData, verbose=FALSE)
+mypcrel <- pcrelate(iterator, pcs = mypcair$vectors[, seq(argv$n_pcs)],
+                    training.set = mypcair$unrels)
+pcrelate_matrix <- pcrelateToMatrix(mypcrel, scaleKin=2, thresh = kin.thresh)
+
+pca <- pcair(seqData, kinobj = pcrelate_matrix, kin.thresh = kin.thresh,
+             divobj = kingMat, snp.include = variant_id,
+             sample.include = sample_id)
+
+resetIterator(iterator, verbose = TRUE)
+
+print("2nd iteration PC-relate")
+iterator <- SeqVarBlockIterator(seqData, verbose = FALSE)
+pcrel2 <- pcrelate(iterator, pcs = pca$vectors[, seq(argv$n_pcs)],
+                   training.set = pca$unrels)
+
+pcrelate_matrix2 <- pcrelateToMatrix(pcrel2, scaleKin = 2, thresh = kin.thresh)
+saveRDS(pca$vectors, paste0(out_prefix, "pcs.rds"))
```

Comment:
It would be useful to save the entire PCA object (maybe in addition to the matrix of eigenvectors by itself). We might want to use the vectors `pca$rels` and `pca$unrels` in future, and `pca$varprop` can be used to plot percent variance explained by each PC (e.g., https://github.com/UW-GAC/analysis_pipeline/blob/master/R/pca_plots.R#L35)